src/main/java/com/aws/greengrass/shadowmanager/sync/CloudDataClient.java

```diff
@@ -145,7 +145,7 @@ public synchronized void updateSubscriptions(Set<Pair<String, String>> shadowSet
      * @param updateTopics Set of update shadow topics to subscribe to
      * @param deleteTopics Set of delete shadow topics to subscribe to
      */
-    private synchronized void updateSubscriptions(Set<String> updateTopics, Set<String> deleteTopics) {
+    private void updateSubscriptions(Set<String> updateTopics, Set<String> deleteTopics) {
```

Comment:
needs a little more refactoring, `subscribedUpdateShadowTopics` and `subscribedDeleteShadowTopics` need synchronization now, maybe they can be passed in as params to this method or something

Reply:
and with this i don't believe we'd need `handleAsync` in the onConnectionResumed anymore

Reply:
But, we are doing some work on mqtt event loop thread which is prone to deadlocks/blocking conditions due to some future changes in any part of the shadow manager code.

Reply:
Both `unsubscribeToShadows` and `subscribeToShadows` are private methods that are `synchronized`. This is still good as they are not blocking like `updateSubscriptions` by retrying indefinitely. Also, `updateSubscriptions` is called from a method that is `synchronized`.

Reply:
> But, we are doing some work on mqtt event loop thread which is prone to deadlocks/blocking conditions due to some future changes in any part of the shadow manager code.   

Which is exactly why we should err towards reducing async/locks where possible, and why with future changes we need to remain vigilant. I'm worried that making this async will cover up more issues.

Reply:
The MQTT client now offers async APIs, since Nucleus 2.10 :)

Reply:
> Both `unsubscribeToShadows` and `subscribeToShadows` are private methods that are `synchronized`. This is still good as they are not blocking like `updateSubscriptions` by retrying indefinitely. Also, `updateSubscriptions` is called from a method that is `synchronized`. 

Right, the issue is that internal updateSubscriptions is sharing the lock with the other methods that don't block